README.md
# commons
一些公用的功能

### commons-util
使用方法：
```xml
<dependency>
    <groupId>cn.kanejin.commons</groupId>
    <artifactId>commons-util</artifactId>
    <version>1.0.2</version>
</dependency>
```


commons-util/CHANGELOG.md
#### 1.0.3 (20170824)
1. 整合字符串相关方法

#### 1.0.2 (20170705)
1. 优化生成随机纯数字的字符串方法
1. 优化日期的部分方法
1. 添加字符串数组的工具类

---
#### 1.0.1 (20170702)
1. 添加String相关的常用方法
1. 添加日期相关的常用方法
1. 添加数字相关的常用方法
1. 添加遮蔽字符串相关的常用方法

---
#### 1.0.0 (20170702)
1. 首次发布


commons-util/src/main/java/cn/kanejin/commons/util/MD5Utils.java
package cn.kanejin.commons.util;import java.io.UnsupportedEncodingException;import java.security.MessageDigest;import java.security.NoSuchAlgorithmException;/** * MD5加密算法 */public class MD5Utils {	/**	 * 对字符串进行MD5加密	 * 	 * @param rawText 明文	 * 	 * @return 密文	 */	public static String md5(String rawText) {		try {			return md5(rawText.getBytes("utf-8"));		} catch (UnsupportedEncodingException e) {			throw new IllegalStateException(					"System doesn't support your  EncodingException.");		}	}	/**	 * 对二进制码进行MD5加密	 *	 * @param rawData 明文二进制码	 *	 * @return	 */	public static String md5(byte[] rawData) {		return new String(encodeHex(encryt(rawData)));	}	private static byte[] encryt(byte[] rawData) {		MessageDigest msgDigest = null;		try {			msgDigest = MessageDigest.getInstance("MD5");		} catch (NoSuchAlgorithmException e) {			throw new IllegalStateException(					"System doesn't support MD5 algorithm.");		}		msgDigest.update(rawData);		return msgDigest.digest();	}	// Used building output as Hex	private static final char[] DIGITS = "0123456789abcdef".toCharArray();	private static char[] encodeHex(byte[] data) {		int l = data.length;		char[] out = new char[l << 1];		// two characters form the hex value.		for (int i = 0, j = 0; i < l; i++) {			out[j++] = DIGITS[(0xF0 & data[i]) >>> 4];			out[j++] = DIGITS[0x0F & data[i]];		}		return out;	}}

commons-util/src/main/java/cn/kanejin/commons/util/DateUtils.java
package cn.kanejin.commons.util;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;

import static cn.kanejin.commons.util.StringUtils.isEmpty;
import static java.util.Calendar.*;

/**
 * 日期相关的方法
 *
 * @author Kane Jin
 */
public class DateUtils {

    /**
     * 格式化日期
     *
     * @param date 日期
     * @param pattern 格式
     * @return 格式化后的日期字符串
     */
    public static String formatDate(Date date, String pattern) {
        if (date == null)
            return "";

        return new SimpleDateFormat(pattern).format(date);
    }

    /**
     * 格式化日期
     *
     * @param time 毫秒数(ms) since 1970/01/01 00:00:00 GMT.
     * @param pattern 格式
     * @return 格式化后的日期字符串
     */
    public static String formatDate(Long time, String pattern) {
        if (time == null)
            return "";

        return formatDate(new Date(time), pattern);
    }

    /**
     * 解析日期字符串
     *
     * @param dateString 日期字符串
     * @param pattern 日期格式
     * @return 解析后的日期
     */
    public static Date parseDate(String dateString, String pattern) {
        if (isEmpty(dateString))
            return null;

        try {
            return new SimpleDateFormat(pattern).parse(dateString);
        } catch (ParseException e) {
            return null;
        }
    }



    /**
     * 判断两个日期是否是同一天
     *
     * @param d1 日期1
     * @param d2 日期2
     * @return 如果两个日期是同一天，返回true，否则返回false
     */
    public static boolean isSameDay(Date d1, Date d2) {
        if (d1 == null || d2 == null)
            return false;

        DateFormat df = new SimpleDateFormat("yyyyMMdd");
        return df.format(d1).equals(df.format(d2));
    }


    /**
     * 获取指定日期一天的开始
     *
     * 例如：
     * "2016/10/20 18:36:58.103"的开始是"2016/10/20 00:00:00.000"
     *
     * @param date 指定日期
     * @return 一天的开始
     */
    public static Date beginOfDate(Date date) {
        if (date == null)
            return null;

        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        startOfDay(cal);

        return cal.getTime();
    }

    /**
     * 获取指定日期一天的开始
     *
     * 例如：
     * "2016/10/20 18:36:58.103"的开始是"2016/10/20 00:00:00.000"
     *
     * @param dateString 指定日期字符串
     * @param pattern 日期格式
     * @return 一天的开始
     */
    public static Date beginOfDate(String dateString, String pattern) {
        if (isEmpty(dateString))
            return null;

        return beginOfDate(parseDate(dateString, pattern));
    }


    /**
     * 获取指定日期一天的结束
     *
     * 例如：
     * "2016/10/20 18:36:58.103"的开始是"2016/10/20 23:59:59.999"
     *
     * @param date 指定日期
     * @return 一天的结束
     */
    public static Date endOfDate(Date date) {
        if (date == null)
            return null;

        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        endOfDay(cal);

        return cal.getTime();
    }

    /**
     * 获取指定日期一天的结束
     *
     * 例如：
     * "2016/10/20 18:36:58.103"的开始是"2016/10/20 23:59:59.999"
     *
     * @param dateString 指定日期字符串
     * @param pattern 日期格式
     * @return 一天的结束
     */
    public static Date endOfDate(String dateString, String pattern) {
        if (isEmpty(dateString))
            return null;

        return endOfDate(parseDate(dateString, pattern));
    }


    /**
     * 获取今天的开始和结束
     *
     * @return 长度为2的日期数组，[0]是今天的开始，[1]是今天的结束
     */
    public static Date[] startAndEndOfToday() {
        Date today = new Date();

        return new Date[]{beginOfDate(today), endOfDate(today)};
    }


    /**
     * 获取指定日期所在星期的开始和结束
     *
     * @param date 指定日期
     * @return 长度为2的日期数组，[0]是星期的开始，[1]是星期的结束
     */
    public static Date[] startAndEndOfWeek(Date date) {

        Date[] result = new Date[2];

        Calendar cal = Calendar.getInstance();
        cal.setTime(date);

        int dayOfWeek = cal.get(DAY_OF_WEEK) - 1;
        if (dayOfWeek == 0)
            dayOfWeek = 7;

        cal.add(DATE, 1 - dayOfWeek);
        startOfDay(cal);
        result[0] = cal.getTime();

        cal.add(DATE, 6);
        endOfDay(cal);
        result[1] = cal.getTime();

        return result;
    }

    /**
     * 获取当前星期的开始和结束
     *
     * @return 长度为2的日期数组，[0]是星期的开始，[1]是星期的结束
     */
    public static Date[] startAndEndOfCurrentWeek() {
        return startAndEndOfWeek(new Date());
    }

    /**
     * 获取指定日期所在月的开始和结束
     *
     * @param date 指定日期
     * @return 长度为2的日期数组，[0]是月的开始，[1]是月的结束
     */
    public static Date[] startAndEndOfMonth(Date date) {
        Date[] result = new Date[2];

        Calendar cal = Calendar.getInstance();
        cal.setTime(date);

        cal.set(DATE, 1);
        startOfDay(cal);
        result[0] = cal.getTime();

        cal.add(MONTH, 1);
        cal.add(DATE, -1);
        endOfDay(cal);
        result[1] = cal.getTime();

        return result;
    }

    /**
     * 获取当前月的开始和结束
     *
     * @return 长度为2的日期数组，[0]是月的开始，[1]是月的结束
     */
    public static Date[] startAndEndOfCurrentMonth() {
        return startAndEndOfMonth(new Date());
    }


    private static int[] QUARTER_FIRST_MONTH = {JANUARY, APRIL, JULY, OCTOBER};


    /**
     * 获取指定日期所在季度的开始和结束
     *
     * @param date 指定日期
     * @return 长度为2的日期数组，[0]是季度的开始，[1]是季度的结束
     */
    public static Date[] startAndEndOfQuarter(Date date) {
        Date[] result = new Date[2];

        Calendar cal = Calendar.getInstance();
        cal.setTime(date);

        cal.set(MONTH, QUARTER_FIRST_MONTH[cal.get(MONTH) / 3]);
        cal.set(DATE, 1);
        startOfDay(cal);
        result[0] = cal.getTime();

        cal.add(MONTH, 3);
        cal.add(DATE, -1);
        endOfDay(cal);
        result[1] = cal.getTime();

        return result;
    }

    /**
     * 获取当前季度的开始和结束
     *
     * @return 长度为2的日期数组，[0]是季度的开始，[1]是季度的结束
     */
    public static Date[] startAndEndOfCurrentQuarter() {
        return startAndEndOfQuarter(new Date());
    }


    /**
     * 获取指定日期所在年的开始和结束
     *
     * @param date 指定日期
     * @return 长度为2的日期数组，[0]是年的开始，[1]是年的结束
     */
    public static Date[] startAndEndOfYear(Date date) {
        Date[] result = new Date[2];

        Calendar cal = Calendar.getInstance();
        cal.setTime(date);

        cal.set(MONTH, JANUARY);
        cal.set(DATE, 1);
        startOfDay(cal);
        result[0] = cal.getTime();

        cal.add(YEAR, 1);
        cal.add(DATE, -1);
        endOfDay(cal);
        result[1] = cal.getTime();

        return result;

    }

    /**
     * 获取当前年的开始和结束
     *
     * @return 长度为2的日期数组，[0]是年的开始，[1]是年的结束
     */
    public static Date[] startAndEndOfCurrentYear() {
        return startAndEndOfYear(new Date());
    }

    private static void startOfDay(Calendar cal) {
        cal.set(HOUR_OF_DAY, 0);
        cal.set(MINUTE, 0);
        cal.set(SECOND, 0);
        cal.set(MILLISECOND, 0);
    }

    private static void endOfDay(Calendar cal) {
        cal.set(HOUR_OF_DAY, 23);
        cal.set(MINUTE, 59);
        cal.set(SECOND, 59);
        cal.set(MILLISECOND, 999);
    }
}


commons-util/src/main/java/cn/kanejin/commons/util/NumberUtils.java
package cn.kanejin.commons.util;

import java.math.BigDecimal;

import static cn.kanejin.commons.util.StringUtils.isEmpty;

/**
 * 数字相关的方法
 *
 * @author Kane Jin
 */
public class NumberUtils {

    /**
     * 判断字符串是否是数字
     *
     * @param str 字符串
     * @return 如果字符串全是数字，则返回true
     */
    public static boolean isInteger(String str) {
        if (isEmpty(str))
            return false;
        return str.matches("^\\d+$");
    }

    /**
     * 判断字符串是否是小数
     *
     * @param str 字符串
     * @return 如果字符串全是数字，则返回true
     */
    public static boolean isDecimal(String str) {
        if (isEmpty(str))
            return false;
        return str.matches("^\\d+(\\.\\d+)?$");
    }

    /**
     * 把字符串转成Integer
     *
     * @param str 字符串
     * @param defaultValue 默认值
     * @return 转后的Integer
     */
    public static Integer toInt(String str, Integer defaultValue) {
        if (isEmpty(str))
            return defaultValue;

        try {
            return Integer.parseInt(str);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    /**
     * 把字符串转成Long
     *
     * @param str 字符串
     * @param defaultValue 默认值
     * @return 转后的Long
     */
    public static Long toLong(String str, Long defaultValue) {
        if (isEmpty(str))
            return defaultValue;

        try {
            return Long.parseLong(str);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    /**
     * 把字符串转成Float
     *
     * @param str 字符串
     * @param defaultValue 默认值
     * @return 转后的Float
     */
    public static Float toFloat(String str, float defaultValue) {
        if (isEmpty(str))
            return defaultValue;

        try {
            return Float.parseFloat(str);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    /**
     * 把字符串转成Double
     *
     * @param str 字符串
     * @param defaultValue 默认值
     * @return 转后的Double
     */
    public static Double toDouble(String str, Double defaultValue) {
        if (isEmpty(str))
            return defaultValue;

        try {
            return Double.parseDouble(str);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    /**
     * 把字符串转成BigDecimal
     *
     * @param str 字符串
     * @param defaultValue 默认值
     * @return 转后的BigDecimal
     */
    public static BigDecimal toBigDecimal(String str, BigDecimal defaultValue) {
        if (isEmpty(str))
            return defaultValue;

        try {
            return new BigDecimal(str);
        } catch (NumberFormatException e) {
            return defaultValue;
        }
    }

    /**
     * 比较两个BigDecimal数值，取其中小的数值
     *
     * @param a 数值a
     * @param b 数值b
     * @return 其中小的数值
     */
    public static BigDecimal min(BigDecimal a, BigDecimal b) {
        return a.compareTo(b) > 0 ? b : a;
    }

    /**
     * 比较两个BigDecimal数值，取其中大的数值
     *
     * @param a 数值a
     * @param b 数值b
     * @return 其中大的数值
     */
    public static BigDecimal max(BigDecimal a, BigDecimal b) {
        return a.compareTo(b) > 0 ? a : b;
    }

}


commons-util/src/main/java/cn/kanejin/commons/util/StringMaskUtils.java
package cn.kanejin.commons.util;

import static cn.kanejin.commons.util.StringUtils.isEmpty;

/**
 * 遮蔽敏感字符串相关的方法
 *
 * @author Kane Jin
 */
public class StringMaskUtils {
    /**
     * 遮蔽敏感字符串
     *
     * 遮蔽字符串中间一半长度的字符，例如：
     * "1rwLQAFbHSh9bZ5GrRwUmKSk9cLft8D1b"遮蔽之后为"1rwLQAFb****************9cLft8D1b"
     *
     * @param str 字符串
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str) {
        return maskString(str, false);
    }

    /**
     * 按指定的位置和长度遮蔽敏感字符串
     *
     * 例如：
     * maskString("13901270809", 3, 4)的返回值为"139****0809"
     *
     * @param str 字符串
     * @param start 开始位置坐标
     * @param length 遮蔽长度
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str, int start, int length) {
        return maskString(str, start, length, false);
    }

    /**
     * 遮蔽敏感字符串，并按要求缩短字符串
     *
     * 遮蔽字符串中间一半长度的字符，如果缩短的话，则只保留前后4位字符，例如：
     * "1rwLQAFbHSh9bZ5GrRwUmKSk9cLft8D1b"遮蔽之后为
     * 未缩短："1rwLQAFb****************9cLft8D1b"
     * 缩短：1rwL**8D1b
     *
     * @param str 字符串
     * @param shorten 是否缩短字符串
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str, boolean shorten) {
        if (isEmpty(str))
            return "";

        int maskLength = 0;

        if (shorten && str.length() > 10) {
            maskLength = Math.max(str.length() - 8, 1);
        } else  {
            maskLength = Math.max(str.length() / 2, 1);
        }

        int startIndex = (str.length() - maskLength) / 2;

        return maskString(str, startIndex, maskLength, shorten);
    }

    /**
     * 按指定的位置和长度遮蔽敏感字符串
     *
     * 如果缩短为true，则遮蔽的部分只用两个**代替，例如：
     * <code>maskString("1rwLQAFbHSh9bZ5GrRwUmKSk9cL", 5, 6, true)</code>
     * 的返回值为<code>"1rwLQ**9bZ5GrRwUmKSk9cL"</code>
     *
     * @param str 字符串
     * @param start 开始位置坐标
     * @param length 遮蔽长度
     * @param shorten 是否缩短字符串
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str, int start, int length, boolean shorten) {
        if (isEmpty(str))
            return "";

        if (start + length >= str.length()) {
            start = 0;
            length = str.length();
        }

        String mask = "";
        if (shorten && length > 2) {
            mask = "**";
        } else {
            for (int i = 0; i < length; i++)
                mask += "*";
        }

        return str.substring(0, start) + mask + str.substring(start + length);
    }

    /**
     * 遮蔽手机号码
     *
     * 等同于<code>maskString(mobile, 3, 4)</code>
     *
     * @param mobile 手机号码
     * @return 遮蔽后的手机号码
     */
    public static String maskMobile(String mobile) {
        return maskString(mobile, 3, 4);

    }

    /**
     * 遮蔽邮箱
     *
     * 邮箱@后面的域名不遮蔽，只遮蔽@之前的用户名，遮蔽用户名中间一半长度的字符，例如：
     * <code>kanejin.cn@gmail.com</code>遮蔽后为<code>ka*****.cn@gmail.com</code>
     *
     * @param email 邮箱
     * @return 遮蔽后的邮箱
     */
    public static String maskEmail(String email) {
        if (isEmpty(email))
            return email;

        if (!email.contains("@"))
            return maskString(email);

        String name = email.substring(0, email.indexOf("@"));

        return maskString(name) + email.substring(email.indexOf("@"));
    }
}


commons-util/src/main/java/cn/kanejin/commons/util/StringArrayUtils.java
package cn.kanejin.commons.util;

import static cn.kanejin.commons.util.StringUtils.isBlank;

/**
 * 字符串数组相关的方法
 *
 * @author Kane Jin
 */
public class StringArrayUtils {
    /**
     * 把字符串数组用分割符串连起来
     *
     * @param array 字符串数组
     * @param separator 分割符
     * @return 串连起来的字符串
     */
    public static String join(String[] array, String separator) {
        if (array == null)
            return "";
        if (separator == null)
            separator = "";

        StringBuilder buf = new StringBuilder();
        for (int i = 0; i < array.length; i++) {
            buf.append(array[i]);
            if (i >= array.length - 1)
                continue;
            buf.append(separator);
        }

        return buf.toString();
    }

    /**
     * 拼接路径
     * <p>
     * 把一些路径拼接起来，比如：
     * 把"http://www.demo.com/", "/user", "/get"
     * 拼接成"http://www.demo.com/user/get"
     *
     * @param pieces 需要拼接的路径
     * @return
     */
    public static String joinPath(String... pieces) {
        if (pieces == null || pieces.length <= 0)
            return "";

        String result = null;
        for (int i = 0; i < pieces.length; i++) {
            if (isBlank(pieces[i]))
                continue;

            if (isBlank(result)) {
                result = pieces[i];
                continue;
            }

            if (result.endsWith("/"))
                if (pieces[i].startsWith("/"))
                    result += pieces[i].substring(1);
                else
                    result += pieces[i];
            else
            if (pieces[i].startsWith("/"))
                result += pieces[i];
            else
                result += "/" + pieces[i];
        }

        return result;
    }
}


commons-util/src/main/java/cn/kanejin/commons/util/StringUtils.java
package cn.kanejin.commons.util;

import java.util.Random;

/**
 * 字符串相关的方法
 *
 * @author Kane Jin
 */
public class StringUtils {
    /**
     * 判断字符串是否为空
     *
     * @param str 字符串
     * @return 如果字符串为null或为空("")，返回true，否则返回false
     */
    public static boolean isEmpty(String str) {
        return str == null || str.isEmpty();
    }

    /**
     * 判断字符串是否不为空
     *
     * @param str 字符串
     * @return 如果字符串为null或为空("")，返回false，否则返回true
     */
    public static boolean isNotEmpty(String str) {
        return !isEmpty(str);
    }

    /**
     * 判断字符串是否为空或仅含有空格
     *
     * @param str 字符串
     * @return 如果字符串为null或为空("")或内容全是空格(" ")，返回true，否则返回false
     */
    public static boolean isBlank(String str) {
        return str == null
            || str.isEmpty()
            || str.trim().isEmpty();
    }

    /**
     * 判断字符串是否不为空而且含有非空格字符
     *
     * @param str 字符串
     * @return 如果字符串为null或为空("")或内容全是空格(" ")，返回false，否则返回true
     */
    public static boolean isNotBlank(String str) {
        return !isBlank(str);
    }

    /**
     * 把为null的字符串转化成空字符串（""）
     *
     * @param str 字符串
     * @return 如果字符串不为null则返回原字符串，否则返回空（""）
     */
    public static String nullToEmpty(String str) {
        return str == null ? "" : str;
    }

    private static final char[] CHARS =
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".toCharArray();
    private static final int CHARS_LENGTH = CHARS.length;
    private static final Random RANDOM = new Random();

    /**
     * 生成一个指定长度的随机字符串（a-zA-Z0-9）
     *
     * @param length 字符串的长度
     * @return 一个指定长度的随机字符串
     */
    public static String random(int length) {
        if (length <= 0)
            throw new IllegalArgumentException("Length must be greater than 0 : " + length);
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < length; i++) {
            sb.append(CHARS[RANDOM.nextInt(CHARS_LENGTH)]);
        }
        return sb.toString();
    }

    /**
     * 生成一个指定长度的随机字符串（a-zA-Z0-9）
     *
     * @param length 字符串的长度
     * @param onlyNumber 是否是全数字，当为true时，只生成全数字的字符串，而且首字符不为'0'
     *
     * @return
     */
    public static String random(int length, boolean onlyNumber) {
        if (!onlyNumber)
            return random(length);

        if (length <= 0)
            throw new IllegalArgumentException("Argument length must be greater than 0 : " + length);
        StringBuilder sb = new StringBuilder();

        for (;;) {
            int i = RANDOM.nextInt(10);

            if (i > 0) {
                sb.append(i);
                break;
            }
        }

        for (int i = 1; i < length; i++) {
            sb.append(RANDOM.nextInt(10));
        }

        return sb.toString();
    }


    /**
     * 遮蔽敏感字符串
     *
     * 遮蔽字符串中间一半长度的字符，例如：
     * "1rwLQAFbHSh9bZ5GrRwUmKSk9cLft8D1b"遮蔽之后为"1rwLQAFb****************9cLft8D1b"
     *
     * @param str 字符串
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str) {
        return maskString(str, false);
    }

    /**
     * 按指定的位置和长度遮蔽敏感字符串
     *
     * 例如：
     * maskString("13901270809", 3, 4)的返回值为"139****0809"
     *
     * @param str 字符串
     * @param start 开始位置坐标
     * @param length 遮蔽长度
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str, int start, int length) {
        return maskString(str, start, length, false);
    }

    /**
     * 遮蔽敏感字符串，并按要求缩短字符串
     *
     * 遮蔽字符串中间一半长度的字符，如果缩短的话，则只保留前后4位字符，例如：
     * "1rwLQAFbHSh9bZ5GrRwUmKSk9cLft8D1b"遮蔽之后为
     * 未缩短："1rwLQAFb****************9cLft8D1b"
     * 缩短：1rwL**8D1b
     *
     * @param str 字符串
     * @param shorten 是否缩短字符串
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str, boolean shorten) {
        if (isEmpty(str))
            return "";

        int maskLength = 0;

        if (shorten && str.length() > 10) {
            maskLength = Math.max(str.length() - 8, 1);
        } else  {
            maskLength = Math.max(str.length() / 2, 1);
        }

        int startIndex = (str.length() - maskLength) / 2;

        return maskString(str, startIndex, maskLength, shorten);
    }

    /**
     * 按指定的位置和长度遮蔽敏感字符串
     *
     * 如果缩短为true，则遮蔽的部分只用两个**代替，例如：
     * <code>maskString("1rwLQAFbHSh9bZ5GrRwUmKSk9cL", 5, 6, true)</code>
     * 的返回值为<code>"1rwLQ**9bZ5GrRwUmKSk9cL"</code>
     *
     * @param str 字符串
     * @param start 开始位置坐标
     * @param length 遮蔽长度
     * @param shorten 是否缩短字符串
     * @return 遮蔽后的字符串
     */
    public static String maskString(String str, int start, int length, boolean shorten) {
        if (isEmpty(str))
            return "";

        if (start + length >= str.length()) {
            start = 0;
            length = str.length();
        }

        String mask = "";
        if (shorten && length > 2) {
            mask = "**";
        } else {
            for (int i = 0; i < length; i++)
                mask += "*";
        }

        return str.substring(0, start) + mask + str.substring(start + length);
    }

    /**
     * 遮蔽手机号码
     *
     * 等同于<code>maskString(mobile, 3, 4)</code>
     *
     * @param mobile 手机号码
     * @return 遮蔽后的手机号码
     */
    public static String maskMobile(String mobile) {
        return maskString(mobile, 3, 4);

    }

    /**
     * 遮蔽邮箱
     *
     * 邮箱@后面的域名不遮蔽，只遮蔽@之前的用户名，遮蔽用户名中间一半长度的字符，例如：
     * <code>kanejin.cn@gmail.com</code>遮蔽后为<code>ka*****.cn@gmail.com</code>
     *
     * @param email 邮箱
     * @return 遮蔽后的邮箱
     */
    public static String maskEmail(String email) {
        if (isEmpty(email))
            return email;

        if (!email.contains("@"))
            return maskString(email);

        String name = email.substring(0, email.indexOf("@"));

        return maskString(name) + email.substring(email.indexOf("@"));
    }


    /**
     * 把字符串数组用分割符串连起来
     *
     * @param array 字符串数组
     * @param separator 分割符
     * @return 串连起来的字符串
     */
    public static String joinArray(String[] array, String separator) {
        if (array == null)
            return "";
        if (separator == null)
            separator = "";

        StringBuilder buf = new StringBuilder();
        for (int i = 0; i < array.length; i++) {
            buf.append(array[i]);
            if (i >= array.length - 1)
                continue;
            buf.append(separator);
        }

        return buf.toString();
    }

    /**
     * 拼接路径
     * <p>
     * 把一些路径拼接起来，比如：
     * 把"http://www.demo.com/", "/user", "/get"
     * 拼接成"http://www.demo.com/user/get"
     *
     * @param pieces 需要拼接的路径
     * @return
     */
    public static String joinPath(String... pieces) {
        if (pieces == null || pieces.length <= 0)
            return "";

        String result = null;
        for (int i = 0; i < pieces.length; i++) {
            if (isBlank(pieces[i]))
                continue;

            if (isBlank(result)) {
                result = pieces[i];
                continue;
            }

            if (result.endsWith("/"))
                if (pieces[i].startsWith("/"))
                    result += pieces[i].substring(1);
                else
                    result += pieces[i];
            else
            if (pieces[i].startsWith("/"))
                result += pieces[i];
            else
                result += "/" + pieces[i];
        }

        return result;
    }
}
